stan/math/gpu/kernels/cholesky_decompose.hpp

```diff
@@ -0,0 +1,81 @@
+#ifndef STAN_MATH_GPU_KERNELS_CHOLESKY_DECOMPOSE_HPP
+#define STAN_MATH_GPU_KERNELS_CHOLESKY_DECOMPOSE_HPP
+#ifdef STAN_OPENCL
+
+#include <stan/math/gpu/kernel_cl.hpp>
+
+namespace stan {
+namespace math {
+namespace opencl_kernels {
+// \cond
+const char *cholesky_decompose_kernel_code = STRINGIFY(
+    // \endcond
+    /**
+     * Calculates the Cholesky Decomposition of a matrix on a GPU
+     *
+     * This kernel is run with threads organized in one dimension and
+     * in a single thread block. The kernel is best suited for
+     * small input matrices as it only utilizes a single streaming
+     * multiprocessor. The kernels is used as a part of a blocked
+     * cholesky decompose.
+     *
+     * @param[in] A The input matrix
+     * @param[in, out] B The result of cholesky decompositon of A.
+     *  It must be set to zeros before execution of this kernel.     *
+     * @param rows The number of rows for A and B.
+     * @note Code is a <code>const char*</code> held in
+     * <code>cholesky_decompose_kernel_code.</code>
+     *  Used in math/gpu/cholesky_decompose.hpp.
+     *  This kernel uses the helper macros available in helpers.cl.
+     *
+     */
+    __kernel void cholesky_decompose(__global double *A, __global double *B,
+                                     int rows) {
+      int local_index = get_local_id(0);
+      // Fill B with zeros
+      // B is square so checking row length is fine for both i and j
+      if (local_index < rows) {
+        for (int k = 0; k < rows; k++) {
+          B(local_index, k) = 0;
+        }
+      }
+      // The following code is the sequential version of the non-inplace
+      // cholesky decomposition. Only the innermost loops are parallelized. The
+      // rows are processed sequentially. This loop process all the rows:
+      for (int j = 0; j < rows; j++) {
+        // First thread calculates the diagonal element
+        if (local_index == 0) {
+          double sum = 0;
+          for (int k = 0; k < j; k++) {
+            sum += B(j, k) * B(j, k);
+          }
+          B(j, j) = sqrt(A(j, j) - sum);
```

Comment:
I tried to get it to resolve by casting the interior to double, and that didn't help, and so I tried turning on [this pragma ](https://www.khronos.org/registry/OpenCL/sdk/1.0/docs/man/xhtml/cl_khr_fp64.html)in the helpers, and that didn't help either... hmmm

Reply:
Can you try using `native_sqrt `? You are trying this on a CPU or GPU?

Reply:
I [ran this](https://stackoverflow.com/a/17881686/212348) and it says my OpenCL implementation doesn't support double precision (which is funny, because I'm just running it on my CPU).

Reply:
Hm, I only encountered this on really old AMD laptop GPUs but thats a long while back.

Reply:
Apparently OpenCL is deprecated on Mac OS X and they may never move past 1.1 (I'm not sure if there will be graphics drivers for Mac that are beyond that - I guess we will see if this problem shows up on the Jenkins Mac Pro.
![image](https://user-images.githubusercontent.com/805356/48982947-e286cf80-f0b6-11e8-8bad-6baa41e12b75.png)

Reply:
Just ran it on the Mac Pro and it seems to work. I guess the graphics card driver has a better implementation or something?

Reply:
Yes, the default CPU OpenCL implementation is weird on MacOS X as far as I have heard. I dont have regular access to a Mac so only I never tested this weirdness. I do know that Apple has a love-hate relationship with its default support. 

GPU drivers have their own implementation and these were never really problematic (nothing new in terms of double support was added in this PR).